### PR TITLE
[WIP] Support shared libraries for x86 and x86_64

### PR DESCRIPTION
**Pull request by partouf**
Needs some more work and testing, but works with gcc and clang.

Combines with https://github.com/mattgodbolt/compiler-explorer-image/pull/246

Also adds LDD as a tool to check the .so paths

Todo:
- [x] OpenSSL
- [x] Clang
- [x] Gcc
- [x] Icc
- [x] Ellcc
  - Uses custom libc libraries, so not possible
- [x] Needs -Wl,-rpath,/opt/compiler-explorer/ellcc-2017-07-16/libecc/lib/x86_64-linux for ldd to work
- [x] Zapcc
- [x] Prepare for MSVC
  - Can link to static lib files, but not execute when redirected to .dll because .dll needs to be copied to executable folder
- [x] C support
- [x] Only pass linking info when filter binary is true
- [x] D support - remove linking from ldc and dmd
- [x] CUDA still works
- [x] Disable for pascal
- [x] Disable for Assemblers
- [x] Disable for Zig

### CODE REVIEW

File: lib/base-compiler.js [hunk 1]
```diff
@@ -202,20 +202,65 @@ class BaseCompiler {
         return options;
     }
 
-    getIncludeArguments(libraries) {
-        const includeFlag = this.compiler.includeFlag || "-I";
+    findLibVersion(selectedLib) {
+        const foundLib = _.find(this.compiler.libs, (o, libId) => {
```
Comment (mattgodbolt) [inline lib/base-compiler.js]: Just a note, this is fine, but can be simplified to `(o, libId) => libId === selectedLib.id` I think (losing the `{}` and `return` and the `()`s :) and below too.

Looks like that was how it was done before: we're all learning how to use modern JS :)

File: lib/map-file.js [hunk 2]
```diff
@@ -435,11 +435,14 @@ class MapFileReader {
...
      * @param {object} address 
      */
-    isWithinAddressSpace(spaces, address) {
+    isWithinAddressSpace(spaces, address, size) {
         for (let idxSpace = 0; idxSpace < spaces.length; ++idxSpace) {
             const spaceAddress = spaces[idxSpace];
 
-            if ((address >= spaceAddress.startAddress) && (address < spaceAddress.endAddress)) {
+            if (((address >= spaceAddress.startAddress) &&
+                 (address < spaceAddress.endAddress)) ||
```
Comment (mattgodbolt) [inline lib/map-file.js]: This check took me a while to double- and triple- check :) But...it is right =) Nice job!

File: lib/map-file.js [hunk 3]
```diff
@@ -435,11 +435,14 @@ class MapFileReader {
      * @param {array of objects} spaces
      * @param {object} address
```
Comment (mattgodbolt) [inline lib/map-file.js]: can you add the param here please

File: lib/map-file.js (hunk 2 above)
Comment (partouf) [reply lib/map-file.js]: should probably have a unittest for it just in case..

**Comment by partouf** [comment]
I broke D's binary mode and exec, need to fix

**Comment by partouf** [comment]
Includes code from a Delphi branch so that I could use the unit tests without having to delete more code before being able to commit :#

**Comment by partouf** [comment]
No idea how D works, so only kept the linking intact for gdc. The rest we can figure out when someone asks for a specific library to link to.

**Pull request merged**